src/toil/jobStores/azureJobStore.py

```diff
@@ -784,26 +825,48 @@ def attempt(last=False):
         try:
             yield
         except retriable_exceptions as e:
-            # Any instance of these exceptions is automatically retriable.
-            if last:
-                raise
+            if time.time() < expiration:
+                # Any instance of these exceptions is automatically retriable.
+                if last:
+                    raise
+                else:
+                    logger.info("Got a retriable exception %s, trying again" % e.__class__.__name__)
             else:
-                logger.info("Got a retriable exception %s, trying again" % e.__class__.__name__)
+                logger.info('Retry timeout expired, giving up.')
+                raise
+
         except Exception as e:
-            # For other exceptions, the retriable_check function determines whether to retry
-            if retriable_check(e):
-                logger.info("Exception %s passed predicate, trying again" % e.__class__.__name__)
+            if time.time() < expiration:
+                # For other exceptions, the retriable_check function determines whether to retry
+                if retriable_check(e):
+                    logger.info("Exception %s passed predicate, trying again" % e.__class__.__name__)
+                else:
+                    raise
             else:
+                logger.info('Retry timeout expired, giving up.')
                 raise
         else:
             go.pop()
 
+    totalTries = numTries
+    totalDelay = 0
+    delay = 0
+
     while go:
-        if num_tries == 1:
-            yield attempt(last=True)
-        else:
-            yield attempt()
+        try:
+            delay = delays[totalTries - numTries]
+        except IndexError:
+            delay = delays[-1]
+        finally:
+            if delay < 0:
```

Comment:
And by assert I mean just check. Don't use the `assert` statement. And the correct exceptions type to raise would be ValueError.